cuML/src/random_projection/rproj.hxx

```diff
@@ -0,0 +1,218 @@
+		auto rng = Random::Rng(params.random_state);
+		math_t scale = 1.0 / sqrt(double(params.n_components));
+		rng.normal(random_matrix->dense_data, len, math_t(0), scale, stream);
+	}
+
+	/**
+	 * @brief generates a sparse random matrix
+	 * @output param random_matrix: the random matrix to be allocated and generated
+	 * @input param params: data structure that includes all the parameters of the model
+	 * @input param stream: cuda stream
+	 */
+	template<typename math_t>
+	void sparse_random_matrix(rand_mat<math_t> *random_matrix, paramsRPROJ& params,
+																		cudaStream_t stream)
+	{
+		if (params.density == 1.0f)
+		{
+			int len = params.n_components * params.n_features;
+			allocate(random_matrix->dense_data, len);
+			auto rng = Random::Rng(params.random_state);
+			math_t scale = 1.0 / sqrt(math_t(params.n_components));
+			rng.scaled_bernoulli(random_matrix->dense_data, len, math_t(0.5), scale, stream);
+		}
+		else
+		{
+			ML::cumlHandle h;
+			auto alloc = h.getHostAllocator();
+
+			double max_total_density = params.density * 1.2;
+			size_t indices_alloc = (params.n_features * params.n_components * max_total_density) * sizeof(int);
+			size_t indptr_alloc = (params.n_components + 1) * sizeof(int);
+			int* indices = (int*)alloc->allocate(indices_alloc, stream);
+			int* indptr = (int*)alloc->allocate(indptr_alloc, stream);
+
+			size_t offset = 0;
+			size_t indices_idx = 0;
+			size_t indptr_idx = 0;
+
+			for (size_t i = 0; i < params.n_components; i++)
+			{
+				int n_nonzero = binomial(params.n_features, params.density);
+				sample_without_replacement(params.n_features, n_nonzero, indices, indices_idx);
+				indptr[indptr_idx] = offset;
+				indptr_idx++;
+				offset += n_nonzero;
+			}
+			indptr[indptr_idx] = offset;
+
+			size_t len = offset;
+			allocate(random_matrix->indices, len);
+			updateDevice(random_matrix->indices, indices, len, stream);
+			alloc->deallocate(indices, indices_alloc, stream);
+
+			len = indptr_idx+1;
+			allocate(random_matrix->indptr, len);
+			updateDevice(random_matrix->indptr, indptr, len, stream);
+			alloc->deallocate(indptr, indptr_alloc, stream);
+
+			len = offset;
+			allocate(random_matrix->sparse_data, len);
+			auto rng = Random::Rng(params.random_state);
+			math_t scale = sqrt(1.0 / params.density) / sqrt(params.n_components);
+			rng.scaled_bernoulli(random_matrix->sparse_data, len, math_t(0.5), scale, stream);
+
+			random_matrix->sparse_data_size = len;
+		}
+	}
+
+	/**
+	 * @brief fits the model by generating appropriate random matrix
+	 * @output param random_matrix: the random matrix to be allocated and generated
+	 * @input param params: data structure that includes all the parameters of the model
+	 */
+	template<typename math_t>
+	void RPROJfit(rand_mat<math_t> *random_matrix, paramsRPROJ* params)
+	{
+		cublasHandle_t cublas_handle;
+		CUBLAS_CHECK(cublasCreate(&cublas_handle));
+
+		cudaStream_t stream;
+		CUDA_CHECK(cudaStreamCreate(&stream));
```

Comment:
Same as above, the stream can be obtained from the `cumlHandle` or be a parameter for lower level functions.